### ISSUE TITLE
Make dedupe easier to use with pandas

### ISSUE THREAD

**Issue by fgregg** [dedupeio/dedupe#680]
Probably won't do this in the core library, but making dedupe work well with pandas could significantly improve adoption in they python data science community. 

I'm not a pandas person, so I'm not sure what it would mean for dedupe to "work well" with pandas. Some notions.

- let people pass in pandas dataframes directly into the dedupe and linking classes
- output a pandas dataframe
- provide an API more similar scikit-learn (I don't know if this is actually helpful to the pandas community, maybe that's a separate issue)
- alternatively to the scikit-learn interface we could use pandas custom accessor interface.

What are other things that would be good for pandas users?

**Comment by fgregg** [comment]
Looks like @Lyonk71 is working on something interesting here: https://github.com/Lyonk71/pandas-dedupe

**Comment by Lyonk71** [comment]
I think that the benefits of connecting Dedupe with Pandas are mostly things like work flow & intuitive design for tabular data. I work with folks in public health who need to deduplicate & link large files. These people tend to be comfortable with SQL, SAS, and R, but experience with Python is not so common. Since these researchers are already comfortable with tables/dataframes & basic programming techniques, I can get them set up with [Ten Minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/10min.html),  and they can start using pandas-dedupe successfully in about 20-30 minutes.

Even for people with more experience, being able to use Dedupe directly on a dataframe takes away from some of the daily grind of analysis work.

Below is a list of convenience/ease-of-use features I’ve been planning for pandas-dedupe:

•	Limit required code to an import statement, loading a dataframe, and a single line of pandas-dedupe code.

•	Incorporation of SQLite3 for working with larger datasets. A user could set a parameter like low_memory=’yes’, and have SQLite run in the background without any setup, or DB knowledge.

•	Enable fuzzy merging with SQL-style joins

•	Allow users to specify which fields they would like to keep canonical versions for. Use canonical version if available, use original value if no canonical version is available. Delete canonical fields before returning to user so that they get the same number of output fields as input fields.

•	If the user specifies a Dedupe type (String, Price, LatLong, etc.), automatically convert the field to be the correct Python data type, include parameter to remove any rows that cannot be converted.

•	For instances where Dedupe is called multiple times in the same program, provide a parameter to read/write specific .db, settings, and training JSON files.

Hope this helps! 

Thanks @adriennefranke for encouraging me to provide documentation for [pandas-dedupe](https://github.com/Lyonk71/pandas-dedupe).

**Comment by fgregg** [comment]
Hi @Lyonk71, thanks for sharing your plans. It's really useful feedback. The auto casting in particular is a very nice idea. 

All these ideas seem very workable, with the possible exception of using sqlite3 for larger dataset. We tried that about five years ago, and sqlite3 really didn't work on mac os x for dedupe (https://github.com/dedupeio/dedupe/issues/72). It's quite possible that the underlying issues with sqlite3's interactions with the os x operating system have been resolved in the intervening years.
